### PR TITLE
Various analyzer fixes

### PR DESCRIPTION
**Pull request by pakrym**
Fixes https://github.com/Azure/azure-sdk-tools/issues/413
Fixes https://github.com/Azure/azure-sdk-tools/issues/396
Fixes https://github.com/Azure/azure-sdk-tools/issues/372

### CODE REVIEW

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs [hunk 1]
```diff
@@ -58,7 +58,7 @@ public override void AnalyzeCore(ISymbolAnalysisContext context)
                         CheckClientMethod(context, syncMember);
                     }
                 }
-            }
-        }
+            }
+        }
```
Comment (AlexanderSher) [inline src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs]: ?

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientConstructorAnalyzer.cs [hunk 2]
```diff
@@ -45,14 +40,19 @@ public override void AnalyzeCore(ISymbolAnalysisContext context)
...
+                    var lastParameter = constructor.Parameters.LastOrDefault();
+
+                    if (IsClientOptionsParameter(lastParameter))
                     {
+                        // Allow optional options parameters
+                        if (lastParameter.IsOptional) continue;
+
                         var nonOptionsMethod = FindMethod(
-                            type.Constructors, constructor.Parameters.RemoveAt(constructor.Parameters.Length - 1));
+                            type.Constructors, constructor.TypeParameters, constructor.Parameters.RemoveAt(constructor.Parameters.Length - 1));
```
Comment (AlexanderSher) [inline src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientConstructorAnalyzer.cs]: Does it create a new array for every constructor?

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs (hunk 1 above)
Comment (pakrym) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs]: Seems like some kind of whitespace normalization, checking.

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientConstructorAnalyzer.cs (hunk 2 above)
Comment (pakrym) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientConstructorAnalyzer.cs]: Yeah, thankfully there aren't that many.

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0001Tests.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ public class Program { }
 
             var diagnostic = Verifier.Diagnostic("AZC0001")
                 .WithMessage("Namespace 'RandomNamespace' shouldn't contain public types. Use one of the following pre-approved namespace groups: " +
-                             "Azure.AI, Azure.Analytics, Azure.Data, Azure.Iot, Azure.Media, Azure.Messaging, Azure.Security, Azure.Storage, Microsoft.Extensions.Azure")
+                             "Azure.AI, Azure.Analytics, Azure.Data, Azure.Iot, Azure.Media, Azure.Messaging, Azure.Security, Azure.Storage, Azure.Identity, Microsoft.Extensions.Azure")
```
Comment (tg-msft) [inline src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0001Tests.cs]: Maybe link to https://azure.github.io/azure-sdk/registered_namespaces.html as well?

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientConstructorAnalyzer.cs [hunk 4]
```diff
@@ -24,18 +24,13 @@ private bool IsClientOptionsParameter(IParameterSymbol symbol)
             {
                 return false;
             }
-            var clientOptionsType = GetOptionsTypeName(symbol.ContainingType);
-            return symbol.Type.Name == clientOptionsType;
+            return symbol.Type.Name.EndsWith(ClientOptionsAnalyzer.ClientOptionsSuffix);
```
Comment (tg-msft) [inline src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientConstructorAnalyzer.cs]: Is it worth checking if the type derives from `ClientOptions`?

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0001Tests.cs (hunk 3 above)
Comment (pakrym) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0001Tests.cs]: Do you feel this rule is useful at all?

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientConstructorAnalyzer.cs (hunk 4 above)
Comment (pakrym) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientConstructorAnalyzer.cs]: I though about it but not deriving from ClientOptions is just not so easy mistake to make.

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0001Tests.cs (hunk 3 above)
Comment (tg-msft) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0001Tests.cs]: Not for our team, but this is a super helpful reminder to new partners just spinning up that they need to work with the Arch Board on their namespace.

**Pull request merged**